### ISSUE TITLE
Hardening Qt and GTK design theming

### ISSUE THREAD

**Issue by GaryElshaw** [transmission/transmission#4420]
### What is the issue?

The more i explore qt on Linux, the more i'm realising some of the behaviour is matching my frustrations with trying to develop gui elements for GTK (#3403). Depending on the theme the user has implemented, their experience can be totally different from the intended design. Somehow, the project needs to harden the design elements so that different themes aren't overriding some of the basic elements.

Case in point. I noticed a left over of a previous pr where the qtr.ico file in /qt is the macOS version that should have been based on this one in /icons

![transmission](https://user-images.githubusercontent.com/69029666/208673900-88ede4dd-83c4-4bb8-829f-ce4f28f1c361.png)

I created a new qtr.ico file for the system tray to emulate the above (pr forthcoming), but nothing changed after a clean build and install. Changing the theme to Adwaita produced another, even more different, system tray icon. Although not in the Adwaita icon base, it is in the Papirus base and appears to be pulled from there.

![Screenshot from 2022-12-20 22-01-36](https://user-images.githubusercontent.com/69029666/208670916-441e82df-9213-4eb1-9512-cbc7ed40e3d5.png)

It leaves the project, i think, two options. Harden the graphic design elements in code to bypass any theming. Or, pushing upstream to theme maintainers of the icons we want seen by users.

### Which application of Transmission?

Qt app on Linux, BSD, etc.

### Which version of Transmission?

Beta 2

**Comment by GaryElshaw** [comment]
And i'm on the opposite side of that fence, because, then what's the point? You might as well ship everything with text buttons and a tooltip for each that says, 'Meh, over to you. Or not ¯\_(ツ)_/¯'. 

I've spent a lot of time in the last few months trying to pull things together so that, at least the divisions of the clients is felt more directly by platform. Also taking into account users and the specifics of the platforms and trying to make things either user-centric (i know what that icon should do by glancing at it). Or platform centric (i know what that button does because they all do that). 

In my experience, and it's because of this project i've spent time in linux distros, we are the only app, i've seen, that has a coloured icon in the tray. And it's not even us putting it there.

**Comment by ckerr** [comment]
This issue seems to be taking a couple of different angles which I have different degrees of agreement / disagreement with.

> Harden the graphic design elements in code to bypass any theming

100% against this. Distros should have final say in what Transmission looks like on their platform, full stop. These are people who are trying to create a consistent experience on their platform and it should be easy for the distro maintainers to make Transmission part of that experience. If that is what this issue is about, LMK so we can move forward.

> I've spent a lot of time in the last few months trying to pull things together so that, at least the divisions of the clients is felt more directly by platform. Also taking into account users and the specifics of the platforms and trying to make things either user-centric (i know what that icon should do by glancing at it). Or platform centric (i know what that button does because they all do that).

I'm trying to meet you halfway here and be open-minded but to be completely honest I don't understand what this paragraph is saying. "dvisions of the clients is felt more directly by platform" seems to be the operative part but I don't know what it means? I'm not trying to troll here.

> In my experience, and it's because of this project i've spent time in linux distros, we are the only app, i've seen, that has a coloured icon in the tray. And it's not even us putting it there.

![tray](https://user-images.githubusercontent.com/70381/209033737-4c7ece91-d576-4069-bc78-f7b118a2d9a6.png)

This is what my tray looks like :man_shrugging:

**Comment by GaryElshaw** [comment]
I think i need to be a little more brain surgeon than orthopaedic surgeon. 

Distros offer a theme (or more commonly, themes) but it's not necessarily what people use to theme. As i mentioned above, the icons adwaita are using aren't in the adwaita theme base, but are being pulled from papirus. Meaning we have one theme that is re-theming Transmission from another. In other words, using Transmission with the adwaita theme isn't even adwaita.

QED: The current qt Transmission icon set we are shipping is more in keeping with adwaita, but adwaita is using papirus.

The alternative i offered was to push the transmission qt icon set to the theme maintainers. I think the adwaita folks would agree that having transmission look adwaita than papirus would be preferable.

>I'm trying to meet you halfway here and be open-minded but to be completely honest I don't understand what this paragraph is saying. "dvisions of the clients is felt more directly by platform" seems to be the operative part but I don't know what it means? I'm not trying to troll here.

Sure thing, i'm just badly trying to say that the application should look and feel like it's coherent with the platform it is on. A bonus, from my point of view, is that the work i'm doing on the qt client is starting to make it look like a cousin of its GTK counterpart.

Interesting array in the tray!

**Comment by Pentaphon** [comment]
> And i'm on the opposite side of that fence, because, then what's the point? You might as well ship everything with text buttons and a tooltip for each that says, 'Meh, over to you. Or not ¯_(ツ)_/¯'.

I agree with you but I find a lot of people on Linux don't use custom icon themes because they just want everything to look the way it comes out of the box so to speak. Therefore, it's still worth making sure that Transmission has modern, complete icons no matter what, because many people will definitely still see them, even if a lot of people use custom icon themes.

**Comment by ckerr** [comment]
Re-reading this issue I'm still not sure what it means. It _sounds_ like this issue is suggesting that we change the code to ensure that our icons are always used. If that's the case I should just close this issue since I _want_ distros to have the final say on what applications on their platform look like.

Being completely honest, I got lost in the part about brain surgeons and orthopaedic surgeons.

So let's take a step back so that you can help walk me through this... leaving all metaphors about hardening & surgery aside, what specific change to the code is this ticket recommending? E.g. what specific files would be changed and for what goal?

**Comment by GaryElshaw** [comment]
> I _want_ distros to have the final say on what applications on their platform look like.

And i'm saying that your awesome and well-intentioned idea that distros would competently do this isn't happening. There are four separate players that effect how the Transmission application appears on the desktop for users, and each can arbitrarily change the entire look and feel of Transmission's dash icon, systray icon, and 'About' icon.

1. The Transmission Project
2. The distro (Transmission is pre-packaged for install)
3. Theme providers
4. Package providers

It isn't being implemented properly when one theme on linux can selectively pull the icons or _an icon_ from another theme and then use 3 separate icon identities as a 'cohesive set'. As i said above, if adwaita is pulling Transmission icon(s) from papirus, the icon theme adwaita users see is actually, or in part, papirus.

Depending on the theme being used, i have seen up to three different icons being used when launching the Qt or GTK client: The macOS dock icon for the dash, the second GTK icon at the top of this issue for the about window, and a third different icon being used in the systray. All unrelated to each other.

I'm not being disingenuous to say that the time and effort others have done designing the client icons and implementing them at the project-level, at the moment with the current code-base, is a waste of design-time if package maintainers don't use them, and then themes over-write them with something else again. Importantly, the icons being used by all of the players above are often being used for **both** the GTK and Qt clients ie, there is no visual identity difference on the desktop between the GTK and Qt client.

That's the problem.

The solution?

There has to be a middle-ground. The project can't change how distros do their work or how packagers do theirs, but the project can bypass themes from the GTK and Qt loop so they at least have a chance of maintaining their separate visual identities on the desktop and have a cohesive look and feel with their sets at least matching.

I'm proposing the code base be changed for the GTK and Qt clients in three ways.

1. A line of code in each of the clients that hard codes what the tray icon is (i've seen some truly weird changes to these made by different themes).
2. I'd like 'our' theme, that's at least coherent, to be immovable, but i get others disagreeing.
3. Coding that ensures the dash icon is replicated to the about window regardless of theme.

**Comment by ckerr** [comment]
> A line of code in each of the clients that hard codes what the tray icon is (i've seen some truly weird changes to these made by different themes).

As mentioned several times above I intentionally want downstream to be able to change the icons so that they have a consistent look with the rest of the platform. It doesn't make sense for us to hardcode a macOS icon on a system that defaults to Adwaita.
